### ISSUE TITLE
Proof creation and verification in Wasm 

### ISSUE THREAD

**Issue by gbotrel** [Consensys-Incorporated/gnark#74]
Identify any issues.

Measure performance of `Groth16` proof creation and/or verification.

**Comment by gbotrel** [comment]
see https://play.gnark.io ; WASM + Groth16 works fine for small circuits, but doesn't scale well at all.

**Comment by ivokub** [comment]
Nice. What version of Tinygo did you use? Have you benchmarked the performance vs native implementation? I think it is even possible to have some good estimates with WASM runtimes in Go a la wasmer-go etc.
Have you tried in browser? How does it benchmark?

I think this is really sought-for feature, but difficult to get just right imo. It is good to see people are working on it independently.

**Comment by p4u** [comment]
The current tinygo develop branch, which fixes an issue with Reflect.

With  llvm 15 installed in your system:

```
git clone https://github.com/tinygo-org/tinygo.git
git switch dev
git submodule update --init
make wasi-libc
go install
```

We will run some benchmarks and post the results.

**Comment by gbotrel** [comment]
🔥 that looks awesome!

re: serialization with cbor, I recall at the time gob was discarded because it didn't handle our largest circuits (100M+ constraints). I think we could use gob for most of the constraint system structure for easier maintainability, and append the constraints themselves in binary encoding at the end. 

finishing a (sigh...) refactoring of the constraint/ package. Agree with the issue wit hintID (have the same one for blueprintID ). 

the sooner we upstream changes that would make gnark/gnark-crypto more wasm/tinygo friendly, the easier it will be to maintain 👍

**Comment by gbotrel** [comment]
@p4u would https://mobile.twitter.com/JohanBrandhorst/status/1645953226494324737 help perf? (doesn't seem to help with binary size in wasm)

**Comment by p4u** [comment]
> @p4u would https://mobile.twitter.com/JohanBrandhorst/status/1645953226494324737 help perf? (doesn't seem to help with binary size in wasm)

I don't think so. Actually we already tested WASI (since tinygo supports it already), see https://hackmd.io/@vocdoni/B1VPA99Z3#WASM-or-WASI

WASI has better support for system calls, but it did not provide any performance advantage. 

IMHO the way to go is tinyGo. The main issue on the WASM scope is the memory allocation (and parallelization but this is not going to be solved anytime soon). TinyGo performs many memory optimizations at compile-time which actually makes a big difference on performance.

**Comment by p4u** [comment]
Tinygo 0.28 and this https://github.com/wasilibs/nottinygc might be a game changer.

**Comment by p4u** [comment]
By the way, currently we are blocked by two things:

1. Hints are still using the reflect mechanism, named hints are already implemented but not used. Maybe tinygo 0.28 might solve that, but it is yet not sure. 
2. Cbor is now serializing an object that cannot be easy replaced by Gob or other tools, that breaks our attempts to test the last develop gnark code.

**Comment by fxamacker** [comment]
Hi, I stumbled across this thread while adding TinyGo support to fxamacker/cbor.

I created branch [fxamacker/cbor-tinygo](https://github.com/fxamacker/cbor/tree/fxamacker/cbor-tinygo) this weekend based on fxamacker/cbor [v2.5.0-beta4](https://github.com/fxamacker/cbor/releases/tag/v2.5.0-beta4).

It passes codec tests when compiled with TinyGo 0.28.1 patched with PR https://github.com/tinygo-org/tinygo/pull/3795.

A tradeoff is the removal of one feature: codec cannot decode CBOR tag to Go interface when compiled with TinyGo. Instead, it will return `UnmarshalError` for now. This tradeoff is caused by TinyGo v0.28.1 not fully implementing `Type.AssignableTo`.

It would be useful to add TinyGo and WASM support to fxamacker/cbor.  Please let me if this works out for you.

**Comment by p4u** [comment]
> Hi, I stumbled across this thread while adding TinyGo support to fxamacker/cbor.
> 
> I created branch [fxamacker/cbor-tinygo](https://github.com/fxamacker/cbor/tree/fxamacker/cbor-tinygo) this weekend based on fxamacker/cbor [v2.5.0-beta4](https://github.com/fxamacker/cbor/releases/tag/v2.5.0-beta4).
> 
> It passes codec tests when compiled with TinyGo 0.28.1 patched with PR [tinygo-org/tinygo#3795](https://github.com/tinygo-org/tinygo/pull/3795).
> 
> A tradeoff is the removal of one feature: codec cannot decode CBOR tag to Go interface when compiled with TinyGo. Instead, it will return `UnmarshalError` for now. This tradeoff is caused by TinyGo v0.28.1 not fully implementing `Type.AssignableTo`.
> 
> It would be useful to add TinyGo and WASM support to fxamacker/cbor. Please let me if this works out for you.

That's great @fxamacker, I will give it a try.

### An update for the efforts of running Gnark on Browser.

I have removed all concurrency of the gnark/gnark-crypto code, so it can be now compiled without a scheduler. 

By tweaking some functions and prioritizing reduction of memory allocation, I have achieved a Groth16 Prover able to solve a circuit of 45k constraints using 1.75 GiB of memory. It takes between 10 and 20 seconds on Android mobile phones with at least 3-4 GiB of RAM.

The big issue and the current stopper is iOS. For some reason, iOS browsers do not allow allocating 2 GiB of memory for WASM execution. I have read somewhere that the limit is 256 MiB (which I hope is not true). 

To this end I am currently trying to use a custom garbage collector for tinyGo that could reduce memory allocation, since the standard "conservative" GC is quite naive, a more optimized GC could make a difference. 

However, we have currently some issues when trying to use it: https://github.com/wasilibs/nottinygc/issues/13

---

If that does not help on iOS devices, my next attempt might be to **write a parser from the Gnark proving key and witness to SnarkJS zkey format**. That would allow building the circuit and generating the witness using Gnark and solving the proof using SnarkJS js/wasm. Some initial work is already done here https://github.com/vocdoni/go-snark/blob/gnark/parsers/gnark/parser.go but just as a non-functional PoC.

We have already tested the same circuit on Circom/SnarkJS and it works properly on iOS and mobile devices. Its code is very optimized for browsers.

IMHO the whole Gnark ecosystem would benefit from such a parser. But the SnarkJS zkey format is very raw and would require some effort... We are considering creating a public bounty for such a task. Let me know if Consensys would be interested in partially funding it. You can also find me at `pau at vocdoni dot io`.

**Comment by p4u** [comment]
Umm, great news! This makes part of the job already: https://github.com/dcbuild3r/ptau-deserializer

**Comment by p4u** [comment]
> So we managed to make it work with a good enough performance!!
> 
> See here the results, conclusions, and benchmarks: https://hackmd.io/@vocdoni/B1VPA99Z3
> 
> Here the code: https://github.com/vocdoni/gnark-prover-tinygo
> 
> Thanks to @ivokub for the help (on this issue https://github.com/ConsenSys/gnark/issues/600)
> 
> We would love to have your feedback and to discuss potential further steps.

See the hackmd link here.

**Comment by fxamacker** [comment]
The build errors in [fxamacker/cbor](https://github.com/fxamacker/cbor) when compiling with TinyGo can be avoided now.

I created a new feature branch (in fxamacker/cbor) tested with tinygo v0.33 and go1.22.  More details at:
- https://github.com/fxamacker/cbor/pull/582 (September 8, 2024)

The feature branch is based on cbor [2.7.0](https://github.com/fxamacker/cbor/releases/tag/v2.7.0), which has memory & speed improvements compared to v2.5.0 used by gnark v0.10.

**Comment by ivokub** [comment]
> Hey @ivokub Did you have a chance to look at this?

Not yet - currently working on https://github.com/Consensys/gnark/pull/1318 and some work for small field solver. But it is on the todo list.

**Comment by Scratch-net** [comment]
Hey @fxamacker I've tried latest tinygo (0.3.5) and cbor from the feature branch.
Good news is that after some minor modifications in Gnark it does compile, but I still get 
`cbor: cannot unmarshal map into Go struct field constraint.System.Blueprints of type constraint.Blueprint` **at runtime** where Blueprint is an [interface](https://github.com/Consensys/gnark/blob/ae6789f16fb069d2cc8f574743dfb93c45f33416/constraint/blueprint.go#L9)
Any hints on how to workaround this?

**Comment by HashMapsData2Value** [comment]
Is there any update to running Gnark in WASM?

**Comment by ivokub** [comment]
Fixed in #1789
